cfme/cloud/keypairs.py

```diff
@@ -50,19 +55,46 @@ def wait_for_delete(self):
 
     def create(self, cancel=False):
         """Create new keypair"""
-        navigate_to(self, 'All')
-        cfg_btn('Add a new Key Pair')
+        navigate_to(self, 'Add')
         fill(keypair_form, {'name': self.name, 'provider': self.provider.name},
              action=keypair_form.save_button)
         if not cancel:
-            flash.assert_message_match('Creating Key Pair {}'.format(self.name))
+            flash.assert_no_errors()
         else:
             flash.assert_message_match('Add of new Key Pair was cancelled by the user')
 
+    def add_tag(self, tag, **kwargs):
+        """Tags the system by given tag"""
+        navigate_to(self, 'Details')
+        keypair_tbl.select_row_by_cells({'Name': self.name})
+        mixins.add_tag(tag, **kwargs)
+
+    def remove_tag(self, tag):
+        """Removes the selected tag off the system"""
+        navigate_to(self, 'Details')
+        keypair_tbl.select_row_by_cells({'Name': self.name})
+        mixins.remove_tag(tag)
+
 
 @navigator.register(KeyPair, 'All')
 class All(CFMENavigateStep):
     prerequisite = NavigateToAttribute('appliance.server', 'LoggedIn')
 
     def step(self):
         self.parent_view.navigation.select('Compute', 'Clouds', 'Key Pairs')
+
+
+@navigator.register(KeyPair, 'Add')
+class Add(CFMENavigateStep):
+    prerequisite = NavigateToSibling('All')
+
+    def step(self):
+        cfg_btn('Add a new Key Pair')
+
+
+@navigator.register(KeyPair, 'Details')
+class Details(CFMENavigateStep):
+    prerequisite = NavigateToSibling('All')
+
```

Comment:
We also don't want to use arguments for step() unless absolutely necessary - the call to tb.refresh should be in the 'resetter' method.  This overrides the default resetter() and is run after step().

You still have the same functionality, because navigate_to() will take a 'use_resetter' boolean. If this is false, the resetter() method that you write (calling tb.refresh) won't be called. Default behavior is to run resetter() after step().